### PR TITLE
Do not store Merkle branches in the wallet.

### LINKED ISSUE #6536: Get rid of Merkle branches in the wallet
They don't currently serve any purpose, except as an expensive sanity check.

However, removing them would result in old clients to no longer treat such transactions as confirmed (until reindexed).

Is this a compatibility break we're willing to make for 0.12?

### PR DESCRIPTION
**Pull request by sipa**
Assume that when a wallet transaction has a valid block hash and transaction position
in it, the transaction is actually there. We're already trusting wallet data in a
much more fundamental way anyway.

To prevent backward compatibility issues, a new record is used for storing the
block locator in the wallet. Old wallets will see a wallet file synchronized up
to the genesis block, and rescan automatically.

Fixes #6536 using a suggestion by @jonasschnelli to retain backward compatibility.

### CODE REVIEW

**Comment by dcousens** [comment]
concept ACK

File: src/primitives/block.cpp [hunk 1]
```diff
@@ -15,7 +15,7 @@ uint256 CBlockHeader::GetHash() const
     return SerializeHash(*this);
 }
 
-uint256 CBlock::BuildMerkleTree(bool* fMutated) const
+uint256 CBlock::ComputeMerkleRoot(bool* fMutated) const
```
Comment (dcousens) [inline src/primitives/block.cpp]: Unrelated to this PR,  but is there any reason we prefer the optional arguments compared to a `tuple` return value?

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -2841,11 +2836,9 @@ int CMerkleTx::GetDepthInMainChainINTERNAL(const CBlockIndex* &pindexRet) const
         return 0;
 
     // Make sure the merkle branch connects to this block
-    if (!fMerkleVerified)
+    if (nIndex == -1)
```
Comment (dcousens) [inline src/wallet/wallet.cpp]: Why is this added?  Its already done above no?
Or is `nIndex` somehow modified in another thread?

File: src/primitives/block.cpp (hunk 1 above)
Comment (laanwj) [inline src/primitives/block.cpp]: I'm not sure this is so much "preferred", it's how it happens to be written between two more-or-less equivalent ways to formulate it.

**Comment by laanwj** [comment]
Concept ACK.

**Comment by jtimon** [comment]
Concept ACK

File: src/primitives/block.cpp (hunk 1 above)
Comment (sipa) [inline src/primitives/block.cpp]: You're relying on an optional optimization allowed by the C++ standard (copy elision) to avoid constructing a tuple as a temporary and copying it to the return location. In C++11 it's better because you can explicitly indicate move semantics.

**Comment by sipa** [comment]
This needs testing :)

**Comment by jonasschnelli** [comment]
Concept ACK. Started testing...

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: Yes. I think this check can be removed completely.

Comment (sipa) [inline src/wallet/wallet.cpp]: Indeed. I missed it was already checked.

**Comment by jonasschnelli** [comment]
**Tested ACK** (lldb stepped; update of "old" wallet.dat works). Played around with some huge regtest wallets/chains. Moved wallet.dat back and forth from master bitcoin-core to master+this PR bitcoin-core.

I think this is a step forward. IMO it's totally sufficient to check if a wtx has a valid block in the active chain and the wtx transaction index matches the position in the block.

Next step could be to simplify the height calculation.
`GetDepthInMainChainINTERNAL()` is very ineffective and holds `cs_main` because it accesses `mapBlockIndex`. Calculating balance, UI listing of transactions will call `GetDepthInMainChainINTERNAL()` for each wallet transaction (! and it even gets called when the credit of a wtx is cached).

 I think a height cache would perform better (together with a invalidating option down to a certain height in case of a reorg).

**Comment by sipa** [comment]
Thanks for testing!

@jonasschnelli No need for a height cache, just a cached chainActive.Tip() inside the wallet suffices (you can do ancestor-of checks without cs_main). I can do that, but it's not for this PR.

**Comment by theuni** [comment]
This will cause the merkle tree to be calculated at least 3x per accepted block, as far as I can see:
ProcessNewBlock -> CheckBlock
ProcessNewBlock -> AcceptBlock -> CheckBlock
ProcessNewBlock -> ActivateBestChain -> ... ConnectBlock -> CheckBlock

Is the calculation time significant enough to try to skip some of that?

**Comment by sipa** [comment]
@theuni Ugh, that's bad. No, it won't make up for that. We should avoid doing those duplicated checks.

**Comment by theuni** [comment]
@sipa If I'm reading correctly, it looks like the CheckBlock() in AcceptBlock() is safe to use !fCheckMerkleRoot since it's only ever called from ProcessNewBlock() where it's already been checked.

Same for the ConnectBlock() in ConnectTip(), since they're already checked before writing to disk. That seems a bit more risky, but surely a bad/corrupt read would be detected long before that point anyway?

**Comment by sipa** [comment]
I guess we want to verify the merkle root any time 1) we receive new block
date 2) read block data back from this.

If the functions who do that always verify, and there is no other way to
construct a full CBlock, I think we're good.

**Comment by sipa** [comment]
@theuni It looks a bit more complicated, as currently we do a CheckBlock on blocks read from disk for validation, which would be harder to do if we want to avoid duplicate checks. I've made a much simpler (and slightly uglier) change for now here, which is to cache the result of CheckBlock in CBlock. That also avoids a few other checks that were being done 3 times before.

**Comment by gmaxwell** [comment]
Where does this stand right now?  Should I spend cycles testing it as is?

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (sipa) [inline src/wallet/wallet.cpp]: Fixed.

**Comment by sipa** [comment]
@gmaxwell Test away.

**Comment by sipa** [comment]
*Ping*

File: src/main.cpp [hunk 3]
```diff
@@ -2573,7 +2576,7 @@ bool CheckBlock(const CBlock& block, CValidationState& state, bool fCheckPOW, bo
     // Check the merkle root.
     if (fCheckMerkleRoot) {
         bool mutated;
-        uint256 hashMerkleRoot2 = block.BuildMerkleTree(&mutated);
+        uint256 hashMerkleRoot2 = block.ComputeMerkleRoot(&mutated);
```
Comment (dcousens) [inline src/main.cpp]: `const`?

**Comment by dcousens** [comment]
re-utACK

**Comment by sipa** [comment]
Rebased.

File: src/main.cpp (hunk 3 above)
Comment (sipa) [inline src/main.cpp]: const what?

File: src/primitives/block.h [hunk 4]
```diff
@@ -78,7 +78,7 @@ class CBlock : public CBlockHeader
     std::vector<CTransaction> vtx;
 
     // memory only
-    mutable std::vector<uint256> vMerkleTree;
+    mutable bool fChecked;
```
Comment (laanwj) [inline src/primitives/block.h]: Ideally we'd have this `fChecked` status (which isn't used in CBlock itself) on an administrative object that wraps a CBlock, instead of on CBlock itself.

Comment (sipa) [inline src/primitives/block.h]: Fully agree, I actually started implementing that as a follow-up, but didn't want to interfere with other refactorings.

Comment (laanwj) [inline src/primitives/block.h]: OK, yes, I'm fine with keeping it like this for this pull

**Comment by laanwj** [comment]
Tested forward/backward compatibility of wallet with this code change.
ACK.
